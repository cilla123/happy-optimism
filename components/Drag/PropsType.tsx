export default interface PropsType {
  onDragStart?: (event?: any, dragState?: Object) => void;
  onDragMove?: (event?: any, dragState?: Object) => void;
  onDragEnd?: (event?: any, dragState?: Object) => void;
  children?: any;
}
