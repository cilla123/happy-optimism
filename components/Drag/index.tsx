import { PureComponent, cloneElement } from 'react';
import PropsType from './PropsType';

export interface DragProps extends PropsType { }

export default class Drag extends PureComponent<DragProps, {}> {

  private dragState: any = {};

  onTouchStart = (event) => {
    const dragState = this.dragState;
    const touch = event.touches[0];

    dragState.startX = touch.pageX;
    dragState.startY = touch.pageY;
    dragState.startTime = new Date();

    const { onDragStart } = this.props;
    if (typeof onDragStart === 'function') {
      onDragStart(event, dragState);
    }
  }

  onTouchMove = (event) => {
    const dragState = this.dragState;
    const touch = event.touches[0];

    const currentX = touch.pageX;
    const currentY = touch.pageY;

    const offsetX = currentX - dragState.startX;
    const offsetY = currentY - dragState.startY;

    const state = {
      ...dragState,
      offsetX,
      offsetY,
      currentX,
      currentY,
    };

    const { onDragMove } = this.props;
    if (typeof onDragMove === 'function' && !onDragMove(event, state)) {
      return;
    }

    this.dragState = state;
  }

  onTouchEnd = (event) => {
    const dragState = this.dragState;
    const { onDragEnd } = this.props;
    if (typeof onDragEnd === 'function') {
      onDragEnd(event, dragState);
    }
    this.dragState = {};
  }

  render(): any {
    const { children } = this.props;
    return cloneElement(children, {
      onTouchStart: this.onTouchStart,
      onTouchMove: this.onTouchMove,
      onTouchEnd: this.onTouchEnd,
    });
  }
}
