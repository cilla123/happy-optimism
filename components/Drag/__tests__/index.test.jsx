import React from 'react';
import { render } from 'enzyme';
import toJson from 'enzyme-to-json';
import Drag from '../index';

describe('Drag', () => {
  it('renders correctly', () => {
    const props = {
      onDragStart: jest.fn(),
      onDragMove: jest.fn(),
      onDragEnd: jest.fn(),
    };
    const wrapper = render(<Drag {...props}><div /></Drag>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
