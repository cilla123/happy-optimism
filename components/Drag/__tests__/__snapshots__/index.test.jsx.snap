// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drag renders correctly 1`] = `<div />`;
